pjlib/src/pj/ssl_sock_schannel.c

```diff
@@ -0,0 +1,1631 @@
+
+    if (sch_ssl.pool) {
+        pj_pool_secure_release(&sch_ssl.pool);
+        pj_caching_pool_destroy(&sch_ssl.cp);
+    }
+
+    pj_bzero(&sch_ssl, sizeof(sch_ssl));
+}
+
+static void sch_inc()
+{
+    if (++sch_ssl.init_cnt == 1 && sch_ssl.pool == NULL) {
+        pj_caching_pool_init(&sch_ssl.cp, NULL, 0);
+        sch_ssl.pool = pj_pool_create(&sch_ssl.cp.factory, "sch%p",
+                                      512, 512, NULL);
+        if (pj_atexit(&sch_deinit) != PJ_SUCCESS) {
+            PJ_LOG(1,(SENDER, "Failed to register atexit() for Schannel."));
+        }
+
+        pj_register_strerror(PJ_SSL_ERRNO_START, PJ_ERRNO_SPACE_SIZE,
+                             &sch_err_print);
+    }
+}
+
+static void sch_dec()
+{
+    pj_assert(sch_ssl.init_cnt > 0);
+    --sch_ssl.init_cnt;
+}
+
+
+/* === SSL socket implementations === */
+
+/* Allocate SSL backend struct */
+static pj_ssl_sock_t *ssl_alloc(pj_pool_t *pool)
+{
+    sch_ssl_sock_t *sch_ssock = NULL;
+
+    sch_inc();
+
+    sch_ssock = (sch_ssl_sock_t*) PJ_POOL_ZALLOC_T(pool, sch_ssl_sock_t);
+    if (!sch_ssock)
+        return NULL;
+
+    SecInvalidateHandle(&sch_ssock->cred_handle);
+    SecInvalidateHandle(&sch_ssock->ctx_handle);
+
+    return &sch_ssock->base;
+}
+
+/* Create and initialize new SSL context and instance */
+static pj_status_t ssl_create(pj_ssl_sock_t *ssock)
+{
+    sch_ssl_sock_t* sch_ssock = (sch_ssl_sock_t*)ssock;
+    pj_pool_factory* pf = ssock->pool->factory;
+    pj_pool_t* pool = ssock->pool;
+    pj_ssize_t read_cap, write_cap;
+    pj_status_t status = PJ_SUCCESS;
+
+    /* Ciphers & curves settings should be set via OS/registry */
+    if (ssock->param.ciphers_num || ssock->param.curves_num) {
+        PJ_LOG(3,(SNAME(ssock), "Ciphers and curves settings are ignored, "
+                  "they should be set via OS/registry"));
+    }
+
+    read_cap  = PJ_MAX(MIN_READ_BUF_CAP,  ssock->param.read_buffer_size);
+    write_cap = PJ_MAX(MIN_WRITE_BUF_CAP, ssock->param.send_buffer_size);
+
+    /* Allocate read buffer */
+    sch_ssock->read_buf_cap = read_cap;
+    sch_ssock->read_buf = (pj_uint8_t*)pj_pool_zalloc(pool, read_cap);
+    if (!sch_ssock->read_buf) {
+        status = PJ_ENOMEM;
+        goto on_return;
+    }
+
+    /* Allocate write buffer */
+    sch_ssock->write_buf_cap = write_cap;
+    sch_ssock->write_buf = (pj_uint8_t*)pj_pool_zalloc(pool, write_cap);
+    if (!sch_ssock->write_buf) {
+        status = PJ_ENOMEM;
+        goto on_return;
+    }
+
+    /* Initialize input circular buffer */
+    status = circ_init(pf, &ssock->circ_buf_input, read_cap);
+    if (status != PJ_SUCCESS)
+        goto on_return;
+
+    /* Initialize output circular buffer */
+    status = circ_init(pf, &ssock->circ_buf_output, write_cap);
+    if (status != PJ_SUCCESS)
+        goto on_return;
+
+    /* Initialize decrypted circular buffer */
+    status = circ_init(pf, &sch_ssock->decrypted_buf, read_cap);
+    if (status != PJ_SUCCESS)
```

Comment:
Can be deleted here?

Reply:
Describe a bit more please, e.g: what can be deleted, what is the potential problem.

Reply:
There's no need to check `status` since the next line is the `on_return` label.

Reply:
Oh, I didn't see the label. IIRC it is on purpose, it relatively gives better readibility & avoid bug when new code is added below that. Hopefully compiler optimization can "delete" it (haven't checked further, but something like [this](https://stackoverflow.com/questions/15301711/stop-goto-optimization-g))?